sky/shell/platform/glfw/message_pump_glfw.cc

```diff
@@ -0,0 +1,87 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "sky/shell/platform/glfw/message_pump_glfw.h"
+
+#include <GLFW/glfw3.h>
+
+#include "base/auto_reset.h"
+#include "base/logging.h"
+#include "base/time/time.h"
+
+namespace sky {
+namespace shell {
+
+MessagePumpGLFW::MessagePumpGLFW() : in_run_(false), should_quit_(false) {
+}
+
+MessagePumpGLFW::~MessagePumpGLFW() {
+}
+
+// static
+scoped_ptr<base::MessagePump> MessagePumpGLFW::Create() {
+  return scoped_ptr<MessagePump>(new MessagePumpGLFW());
+}
+
+void MessagePumpGLFW::Run(Delegate* delegate) {
+  base::AutoReset<bool> auto_reset_keep_running(&should_quit_, false);
+  base::AutoReset<bool> auto_reset_in_run(&in_run_, true);
+
+  for (;;) {
+    bool did_work = delegate->DoWork();
+    if (should_quit_)
+      break;
+
+    did_work |= delegate->DoDelayedWork(&delayed_work_time_);
+    if (should_quit_)
+      break;
+
+    if (did_work)
+      continue;
+
+    did_work = delegate->DoIdleWork();
+    if (should_quit_)
+      break;
+
+    if (did_work)
+      continue;
+
+    if (delayed_work_time_.is_null()) {
+      glfwPollEvents();
+    } else {
+      base::TimeDelta delay = delayed_work_time_ - base::TimeTicks::Now();
+      if (delay > base::TimeDelta()) {
+        glfwWaitEventsTimeout(delay.InSecondsF());
+      } else {
+        // It looks like delayed_work_time_ indicates a time in the past, so we
+        // need to call DoDelayedWork now.
+        delayed_work_time_ = base::TimeTicks();
```

Comment:
Not quite sure what setting this does (if anything). `DoDelayedWork` only uses this as an out parameter. You can probably just continue from here or get rid of the else block entirely.